Infer top-level nullability independently from the type

The following should infer `A?` for `F(x, y)` rather than reporting a warning:
```c#
class A { }
class B : A { }
class Program
{
    static void F<T>(T x, T y)
    {
    }
    static void G(A x, B? y)
    {
        F(x, y); // warning CS8604: possible null 'y' in 'void F<A>(A x, A y)'
    }
}
```

--- a/src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs
+++ b/src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs
@@ -2850,6 +2850,12 @@ OuterBreak:
                     if (!ImplicitConversionExists(source, destination, ref useSiteDiagnostics, conversions.WithNullability(false)))
                     {
                         candidates.Remove(candidate);
+                        if (conversions.IncludeNullability)
+                        {
+                            candidates.Remove(bound);
+                            var newBound = TypeWithAnnotations.Create(bound.Type, bound.NullableAnnotation.MergeNullableAnnotation(candidate.NullableAnnotation, variance));
+                            candidates.Add(newBound, newBound);
+                        }
                     }
                     else if (bound.Equals(candidate, TypeCompareKind.IgnoreDynamicAndTupleNames | TypeCompareKind.IgnoreNullableModifiersForReferenceTypes))
                     {

--- a/src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs
+++ b/src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs
@@ -2811,10 +2811,6 @@ OuterBreak:
             }
         }
 
-        /// <summary>
-        /// Returns false if there was a conflict merging nullability.
-        /// In that case, a warning should be reported by the caller.
-        /// </summary>
         private static void MergeOrRemoveCandidates(
             Dictionary<TypeWithAnnotations, TypeWithAnnotations> candidates,
             HashSet<TypeWithAnnotations> bounds,
@@ -2850,11 +2846,16 @@ OuterBreak:
                     if (!ImplicitConversionExists(source, destination, ref useSiteDiagnostics, conversions.WithNullability(false)))
                     {
                         candidates.Remove(candidate);
-                        if (conversions.IncludeNullability)
+                        if (conversions.IncludeNullability && candidates.TryGetValue(bound, out var oldBound))
                         {
-                            candidates.Remove(bound);
-                            var newBound = TypeWithAnnotations.Create(bound.Type, bound.NullableAnnotation.MergeNullableAnnotation(candidate.NullableAnnotation, variance));
-                            candidates.Add(newBound, newBound);
+                            // merge the nullability from candidate into bound
+                            var oldAnnotation = oldBound.NullableAnnotation;
+                            var newAnnotation = oldAnnotation.MergeNullableAnnotation(candidate.NullableAnnotation, variance);
+                            if (oldAnnotation != newAnnotation)
+                            {
+                                var newBound = TypeWithAnnotations.Create(oldBound.Type, newAnnotation);
+                                candidates[bound] = newBound;
+                            }
                         }
                     }
                     else if (bound.Equals(candidate, TypeCompareKind.IgnoreDynamicAndTupleNames | TypeCompareKind.IgnoreNullableModifiersForReferenceTypes))

--- a/src/Compilers/CSharp/Portable/Symbols/NullableAnnotationExtensions.cs
+++ b/src/Compilers/CSharp/Portable/Symbols/NullableAnnotationExtensions.cs
@@ -1,5 +1,7 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+using Roslyn.Utilities;
+
 namespace Microsoft.CodeAnalysis.CSharp.Symbols
 {
     internal static class NullableAnnotationExtensions
@@ -43,6 +45,20 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
             if (b.IsOblivious())
                 return a;
             return (a < b) ? a : b;
+        }
+
+        /// <summary>
+        /// Merges nullability.
+        /// </summary>
+        public static NullableAnnotation MergeNullableAnnotation(this NullableAnnotation a, NullableAnnotation b, VarianceKind variance)
+        {
+            return variance switch
+            {
+                VarianceKind.In => a.Meet(b),
+                VarianceKind.Out => a.Join(b),
+                VarianceKind.None => a.EnsureCompatible(b),
+                _ => throw ExceptionUtilities.UnexpectedValue(variance)
+            };
         }
     }
 }

--- a/src/Compilers/CSharp/Portable/Symbols/TypeWithAnnotations.cs
+++ b/src/Compilers/CSharp/Portable/Symbols/TypeWithAnnotations.cs
@@ -241,28 +241,10 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
         internal TypeWithAnnotations MergeNullability(TypeWithAnnotations other, VarianceKind variance)
         {
             TypeSymbol typeSymbol = other.Type;
-            NullableAnnotation nullableAnnotation = MergeNullableAnnotation(this.NullableAnnotation, other.NullableAnnotation, variance);
+            NullableAnnotation nullableAnnotation = this.NullableAnnotation.MergeNullableAnnotation(other.NullableAnnotation, variance);
             TypeSymbol type = Type.MergeNullability(typeSymbol, variance);
             Debug.Assert((object)type != null);
             return Create(type, nullableAnnotation, CustomModifiers);
-        }
-
-        /// <summary>
-        /// Merges nullability.
-        /// </summary>
-        private static NullableAnnotation MergeNullableAnnotation(NullableAnnotation a, NullableAnnotation b, VarianceKind variance)
-        {
-            switch (variance)
-            {
-                case VarianceKind.In:
-                    return a.Meet(b);
-                case VarianceKind.Out:
-                    return a.Join(b);
-                case VarianceKind.None:
-                    return a.EnsureCompatible(b);
-                default:
-                    throw ExceptionUtilities.UnexpectedValue(variance);
-            }
         }
 
         public TypeWithAnnotations WithModifiers(ImmutableArray<CustomModifier> customModifiers) =>

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs
@@ -86538,5 +86538,64 @@ class B : A { }
             comp.VerifyDiagnostics();
             comp.VerifyTypes();
         }
+
+        [WorkItem(30987, "https://github.com/dotnet/roslyn/issues/30987")]
+        [Fact]
+        public void TypeInference_LowerBounds_TopLevelNullability_04()
+        {
+            var source =
+@"class D
+{
+    static T F<T>(T x, T y, T z) => x;
+    static void G1(A x, B? y, C z)
+    {
+        F(x, y, z)/*T:A?*/;
+        F(y, z, x)/*T:A?*/;
+        F(z, x, y)/*T:A?*/;
+        F(x, z, y)/*T:A?*/;
+        F(z, y, x)/*T:A?*/;
+        F(y, x, z)/*T:A?*/;
+    }
+}
+class A { }
+class B : A { }
+class C : A { }
+";
+            var comp = CreateCompilation(new[] { source }, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics();
+            comp.VerifyTypes();
+        }
+
+        [WorkItem(30987, "https://github.com/dotnet/roslyn/issues/30987")]
+        [Fact]
+        public void TypeInference_LowerBounds_TopLevelNullability_05()
+        {
+            var source =
+@"class D
+{
+#nullable disable
+    static T F<T>(T x, T y, T z) => x;
+    static void G1(A x, B? y, C z)
+#nullable enable
+    {
+        F(x, y, z)/*T:A?*/;
+        F(y, z, x)/*T:A?*/;
+        F(z, x, y)/*T:A?*/;
+        F(x, z, y)/*T:A?*/;
+        F(z, y, x)/*T:A?*/;
+        F(y, x, z)/*T:A?*/;
+    }
+}
+class A { }
+class B : A { }
+class C : A { }
+";
+            var comp = CreateCompilation(new[] { source }, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (5,26): warning CS8632: The annotation for nullable reference types should only be used in code within a '#nullable' context.
+                //     static void G1(A x, B? y, C z)
+                Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(5, 26));
+            comp.VerifyTypes();
+        }
     }
 }

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs
@@ -86491,5 +86491,52 @@ public class C
                 Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "__refvalue(r, string?)").WithLocation(7, 13)
                 );
         }
+
+        [WorkItem(30987, "https://github.com/dotnet/roslyn/issues/30987")]
+        [Fact]
+        public void TypeInference_LowerBounds_TopLevelNullability_03()
+        {
+            var source =
+@"using System;
+class C
+{
+    static T F<T>(T x, T y) => x;
+    static void G1(A x, B? y)
+    {
+        F(x, x)/*T:A!*/;
+        F(x, y)/*T:A?*/;
+        F(y, x)/*T:A?*/;
+        F(y, y)/*T:B?*/;
+    }
+    static void G2(A x, B? y)
+    {
+        _ = new [] { x, x }[0]/*T:A!*/;
+        _ = new [] { x, y }[0]/*T:A?*/;
+        _ = new [] { y, x }[0]/*T:A?*/;
+        _ = new [] { y, y }[0]/*T:B?*/;
+    }
+    static T M<T>(Func<T> func) => func();
+    static void G3(bool b, A x, B? y)
+    {
+        M(() => { if (b) return x; return x; })/*T:A!*/;
+        M(() => { if (b) return x; return y; })/*T:A?*/;
+        M(() => { if (b) return y; return x; })/*T:A?*/;
+        M(() => { if (b) return y; return y; })/*T:B?*/;
+    }
+    static void G4(bool b, A x, B? y)
+    {
+        _ = (b ? x : x)/*T:A!*/;
+        _ = (b ? x : y)/*T:A?*/;
+        _ = (b ? y : x)/*T:A?*/;
+        _ = (b ? y : y)/*T:B?*/;
+    }
+}
+class A { }
+class B : A { }
+";
+            var comp = CreateCompilation(new[] { source }, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics();
+            comp.VerifyTypes();
+        }
     }
 }

